pkg/sortition/sortition.go

```diff
@@ -0,0 +1,118 @@
+package sortition
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-log"
+)
+
+const (
+	DefaultStatusCheckTick = 10 * time.Second
+)
+
+var logger = log.Logger("keep-sortition")
+
+var errOperatorUnknown = fmt.Errorf("operator not registered for the staking provider, check Threshold dashboard")
+
+// MonitorPool periodically checks the status of the operator in the sortition
+// pool. If the operator is supposed to be in the sortition pool but is not
+// there yet, the function attempts to add the operator to the pool. If the
+// operator is already in the pool and its status is no longer up to date, the
+// function attempts to update the operator's status in the pool.
+func MonitorPool(
+	ctx context.Context,
+	chain Chain,
+	tick time.Duration,
+) error {
+	_, isRegistered, err := chain.OperatorToStakingProvider()
+	if err != nil {
+		return fmt.Errorf("could not resolve staking provider: [%v]", err)
+	}
+
+	if !isRegistered {
+		return errOperatorUnknown
+	}
+
+	err = checkOperatorStatus(chain)
+	if err != nil {
+		logger.Errorf("could not check operator sortition pool status: [%v]", err)
+	}
+
+	ticker := time.NewTicker(tick)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				err = checkOperatorStatus(chain)
+				if err != nil {
+					logger.Errorf("could not check operator sortition pool status: [%v]", err)
+					continue
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func checkOperatorStatus(chain Chain) error {
+	logger.Info("checking sortition pool operator status")
+
+	isOperatorInPool, err := chain.IsOperatorInPool()
+	if err != nil {
+		return err
+	}
+
+	if isOperatorInPool {
+		logger.Info("operator is in the sortition pool")
+	} else {
+		logger.Info("operator is not in the sortition pool")
```

Comment:
Just to make sure we don't `return` here deliberately. We still want to check the operator status even though the operator is not in the pool right?

Reply:
Yes. `IsOperatorUpToDate` not only checks whether the operator that is already in pool has up-to-date weight but also if operator is supposed to be in the sortition pool. That is, if someone is *not* in the sortition pool but their eligible stake is non-zero, `IsOperatorUpToDate` returns false.

Reply:
This flow can be translated to English as:
- Are you in the pool? Yes. Are you up to date?
- Are you in the pool? No. Are you supposed to be in the pool?